File: First_and_Last.js
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Finding the first occurance of the element was a challenge.
//                                           Find the indexes was easy after that.  




/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    
    const BinarySearch = (low, high, left) => {
         while(low < high){
        let mid = Math.trunc((low + high) / 2);
        if(nums[mid] > target || left && target === nums[mid]) {
            high = mid;
            }
             else {
                 low = mid + 1
             }
        }
        return low
    };
    
    const result = [-1,-1];
    const first = BinarySearch(0,nums.length, true);
    if(result >= nums.length || nums[result] != target){
        return result
    }
    
    result[0] = [first];
    result[1] = BinarySearch(0, nums.length, false) - 1;
    return result;
    
    // let len = nums.length;
    // let low = 0, high = len;
    
    
   
};

File: first and last.java
class Solution {

    public int first(int[] nums, int low, int high, int target)

    {

        while (low <= high){

            int mid = low + (high - low)/2;

            if(nums[mid] == target){

                if( mid == low ||  nums[mid - 1] < nums[mid]) return mid;

                // still go left

                else high = mid - 1;

            } else if(nums[mid] > target){

                high = mid - 1;

            } else low = mid + 1; 

        }

        return -1; 

    }


    public int second(int[] nums, int low, int high, int target)

    {

        while (low <= high){

            int mid = low + (high - low)/2;

            if(nums[mid] == target){

                if(mid == high || nums[mid + 1] > target) return mid;

                // still go right

                else low = mid + 1;

            } else if(nums[mid] > target){

                high = mid - 1;

            } else low = mid + 1; 

        }

        return -1; 

    }

    public int[] searchRange(int[] nums, int target) {

        int left = first(nums, 0, nums.length - 1, target);

        int right = second(nums, 0, nums.length - 1, target);

        return new int[] {left,right};

    }

}